genesis/src/lockups.rs

```diff
@@ -0,0 +1,160 @@
+//! lockups generator
+use solana_sdk::{clock::Epoch, epoch_schedule::EpochSchedule, timing::years_as_slots};
+use std::time::Duration;
+
+#[derive(Debug, Default, Clone)]
+pub struct Lockups {
+    /// where in iteration progress I am
+    expiration: usize,
+    /// number of expirations post cliff
+    expirations: usize,
+    /// fractionage offered to f() last time
+    prev_fraction: f64,
+    /// portion of total available at cliff
+    cliff_fraction: f64,
+    /// time of cliff, in epochs
+    cliff_epoch: Epoch,
+    /// number of post-cliff events
+    expiration_fraction: f64,
+    /// time between each post-cliff event
+    expiration_epochs: Epoch,
```

Comment:
It'd be nice to see this calculated from `total_epochs`.  That way it reads "3 year lockup with a 1 year cliff" and not "1 year cliff and then a monthly cliff, such that all tokens end up unlocked after 3 years".

Reply:
cool, sounds like fun

Reply:
```rust
let expiration_epochs = total_epochs * (1.0 - cliff_fraction) / num_expirations;
```

Reply:
ah.  can't do that.  epochs aren't of uniform duration.